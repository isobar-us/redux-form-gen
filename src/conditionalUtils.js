// @flow
import get from 'lodash/get';
import has from 'lodash/has';
import hasIn from 'lodash/hasIn';
import includes from 'lodash/includes';
import omit from 'lodash/omit';
import isEqual from 'lodash/isEqual';
import isNumber from 'lodash/isNumber';
import isNil from 'lodash/isNil';
import isString from 'lodash/isString';
import {isNilOrEmpty, isFieldFilled, isFieldValid} from './validators';

import type {FieldType} from './types';
import type {ConditionalOperators, EvalCondOptions, ConditionalObject} from './conditionalUtils.types';

/*
  Helpers
 */

// used for render lifecycle where we don't care about conditionalValid
export const isCondField = (field: FieldType) =>
  has(field, 'conditionalVisible') || has(field, 'conditionalRequired') || has(field, 'conditionalDisabled');

const getCondValueKey = ({cond, valueKey}) => (has(cond, 'questionId') ? cond.questionId : valueKey);

const getOperator = (options, key) => {
  const {customOperators} = options;
  const operators = {...ops, ...(customOperators || {})};
  const op = get(operators, key);
  if (isNil(op)) {
    console.error(`[FormGenerator] Unknown conditional operator "${key}". Condition:`, options.cond);
  }
  return op;
};

/*
  Conditional Operators
 */
const ops: ConditionalOperators = {
  equals: ({value, param}) => isEqual(value, param),
  // cond
  and: ({value, param, ...options}) =>
    param.reduce((and, subCond) => and && evalCond({...options, cond: subCond}), true),
  or: ({value, param, ...options}) => param.reduce((or, subCond) => or || evalCond({...options, cond: subCond}), false),
  not: ({value, param, ...options}) => !evalCond({...options, cond: param}),
  cond: ({value, param, ...options}) => evalCond({...options, cond: param}),
  compare: (options) => {
    const {value, param, data} = options;

    const conds = Object.keys(omit(param, ['questionId']));
    return conds.length > 0
      ? conds.reduce((result, key) => {
          // will AND all the cond props
          const operator = getOperator(options, key);
          if (isNil(operator)) {
            return true;
          }
          const remoteQuestionId = get(param, key);
          const compareParam = get(
            data,
            getCondValueKey({
              ...options,
              cond: {
                ...(remoteQuestionId && {questionId: remoteQuestionId})
              }
            })
          );
          return result && operator({...options, value, param: compareParam});
        }, true)
      : true;
  },
  // generator hooks
  filled: ({value, param, ...options}) => {
    let fieldFilled = true;
    if (has(options, 'lookupTable')) {
      const {lookupTable} = options;
      const field = get(lookupTable, getCondValueKey(options));
      if (!isNil(field)) {
        // if field is defined in the lookupTable, run isFieldFilled()
        fieldFilled = isFieldFilled({...options, field, fieldOptions: null});
      } else {
        // if field is not defined in the lookupTable, just fall back to isNilOrEmpty
        fieldFilled = !isNilOrEmpty(value);
      }
    } else {
      console.warn('[Form Generator] attempted to check `filled` without a lookupTable. Condition:', options.cond); // TODO finalise this warning
      fieldFilled = !isNilOrEmpty(value);
    }
    return param === true ? fieldFilled : !fieldFilled;
  },
  // TODO experimental operator to check if this or another field is valid...
  // valid: ({value, param, ...options}) => {
  //   const {cond, lookupTable} = options;
  //   let field = options.field;
  //
  //   if (has(cond, 'questionId')) {
  //     if (has(options, 'lookupTable')) {
  //       field = get(lookupTable, cond.questionId);
  //     } else {
  //       console.warn('[Form Generator] attempted to check `valid` without a lookupTable. Condition:', options.cond); // TODO finalise this warning
  //     }
  //   }
  //
  //   const fieldValid = isFieldValid({
  //     ...options,
  //     field,
  //     fieldOptions: null,
  //   });
  //   return param === true ? fieldValid : !fieldValid;
  // },
  // lodash
  includes: ({value, param}) => includes(value, param),
  // comparison
  greaterThan: ({value, param}) => {
    if (isNilOrEmpty(value)) {
      return false;
    }
    const parsedValue = isNumber(value) ? value : parseFloat(value);
    return parsedValue > param;
  },
  lessThan: ({value, param}) => {
    if (isNilOrEmpty(value)) {
      return false;
    }
    const parsedValue = isNumber(value) ? value : parseFloat(value);
    return parsedValue < param;
  },
  greaterThanEqual: ({value, param}) => {
    if (isNilOrEmpty(value)) {
      return false;
    }
    const parsedValue = isNumber(value) ? value : parseFloat(value);
    return parsedValue >= param;
  },
  lessThanEqual: ({value, param}) => {
    if (isNilOrEmpty(value)) {
      return false;
    }
    const parsedValue = isNumber(value) ? value : parseFloat(value);
    return parsedValue <= param;
  },
  // length
  length: ({value, param}) => (isNil(value) ? false : hasIn(value, 'length') ? value.length === param : true),
  minLength: ({value, param}) => (isNil(value) ? false : hasIn(value, 'length') ? value.length >= param : true),
  maxLength: ({value, param}) => (hasIn(value, 'length') ? value.length <= param : true),
  // regex
  pobox: ({value, param}) => (param && isString(value) ? /p.*o.* box/gi.test(value) : true),
  email: ({value, param}) => (param && isString(value) ? /\S+@\S+\.\S+/.test(value) : true),
  regex: ({value, param}) => (isString(value) ? RegExp(param).test(value) : true)
};

/*
  Conditional Operator Aliases
 */
ops.gt = ops.greaterThan;
ops.lt = ops.lessThan;
ops.min = ops.greaterThanEqual;
ops.max = ops.lessThanEqual;
ops.value = ops.equals;

const defaultElseHandler = (options) => ops.filled({...options, param: true});

const defaultValueKey = '_value';

export const evalCond = (options: EvalCondOptions) => {
  options = {
    elseHandler: defaultElseHandler,
    valueKey: defaultValueKey,
    ...options
  };
  const {cond, data, elseHandler} = options;

  // TODO add support for getFieldPath() when doing get(data, path) ?
  const value = get(data, getCondValueKey(options));
  const conds = Object.keys(omit(cond, 'questionId'));

  return conds.length > 0
    ? conds.reduce((result, key) => {
        // will AND all the cond props
        const operator = getOperator(options, key);
        if (isNil(operator)) {
          return true;
        }
        const param = get(cond, key);
        return result && operator({...options, value, param});
      }, true)
    : elseHandler({...options, value});
};

const condValidElseHandler = (options) => {
  const {cond, lookupTable} = options;
  let field = options.field;
  let fieldOptions = null;

  if (has(cond, 'questionId')) {
    if (has(options, 'lookupTable')) {
      field = get(lookupTable, cond.questionId);
      fieldOptions = null;
    } else {
      console.warn('[Form Generator] attempted to check `valid` without a lookupTable. Condition:', options.cond); // TODO finalise this warning
    }
  }

  return isFieldValid({
    ...options,
    field,
    fieldOptions
  });

  // TODO hook into experimental `valid` operator when ready
  // return ops.valid({...options, param: true});
};
export const evalCondValid = (options: EvalCondOptions) => {
  const {data} = options;
  const value = get(data, getCondValueKey(options));

  // can not validate if the field is empty, so return true
  if (isNil(value) || value === '') {
    return true;
  }

  return evalCond({
    elseHandler: condValidElseHandler,
    ...options
  });
};

export default {
  evalCond,
  evalCondValid
};

// TODO consider renaming to getCondDependentFieldNames or something similar
export const condDependentFields = (cond: ConditionalObject) => {
  if (has(cond, 'questionId')) {
    return [cond.questionId];
  } else if (has(cond, 'and') && Array.isArray(cond.and)) {
    return cond.and.reduce((a, subCond) => a.concat(condDependentFields(subCond)), []);
  } else if (has(cond, 'or') && Array.isArray(cond.or)) {
    return cond.or.reduce((a, subCond) => a.concat(condDependentFields(subCond)), []);
  } else if (has(cond, 'not') && cond.not) {
    return condDependentFields(cond.not);
  }
  return [];
};
