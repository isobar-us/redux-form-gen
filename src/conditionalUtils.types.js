import type {FieldType} from './types';
export type ConditionalObject = {
  questionId: string,
  equals?: mixed,
  // cond
  and?: Array<ConditionalObject>,
  or?: Array<ConditionalObject>,
  not?: ConditionalObject,
  // lodash
  filled?: boolean,
  includes?: mixed | Array<mixed>,
  // comparison
  greaterThan?: number,
  lessThan?: number,
  greaterThanEqual?: number,
  lessThanEqual?: number,
  // length
  length: number,
  minLength: number,
  maxLength: number,
  // regex
  pobox: boolean,
  email: boolean,
  regex: boolean
};

export type ConditionalDependency = {
  questionId: string,
  globalScope?: Boolean
};

export type FieldDependencyOptions = {
  field: FieldType,
  parentPath: String
};

export type EvalCondOptions = {
  elseHandler?: Function,
  valueKey?: string,
  cond: ConditionalObject,
  customOperators: CustomOperators // eslint-disable-line no-use-before-define
};

export type CustomOperators = {
  [key: string]: {(options: EvalCondOptions): boolean}
};

export type ConditionalOperatorOptions = {
  value: mixed,
  param: mixed,
  ...EvalCondOptions
};

type ConditionalOperatorNumberParam = {
  (options: {...ConditionalOperatorOptions, value: number, param: number}): boolean
};
type ConditionalOperatorObjectParam = {
  (options: {...ConditionalOperatorOptions, value: Object, param: boolean}): boolean
};

type ConditionalOperator = {(options: ConditionalOperatorOptions): boolean};

export type ConditionalOperators = {
  equals: ConditionalOperator,
  and: {(options: {...ConditionalOperatorOptions, param: Array<ConditionalObject>}): boolean},
  or: {(options: {...ConditionalOperatorOptions, param: Array<ConditionalObject>}): boolean},
  not: {(options: {...ConditionalOperatorOptions, param: ConditionalObject}): boolean},
  filled: ConditionalOperator,
  includes: {
    (options: {...ConditionalOperatorOptions, value: Array<mixed>, param: mixed}): boolean
  },
  greaterThan: ConditionalOperatorNumberParam,
  lessThan: ConditionalOperatorNumberParam,
  greaterThanEqual: ConditionalOperatorNumberParam,
  lessThanEqual: ConditionalOperatorNumberParam,
  length: ConditionalOperatorObjectParam,

  gt?: ConditionalOperatorNumberParam,
  lt?: ConditionalOperatorNumberParam,
  min?: ConditionalOperatorNumberParam,
  max?: ConditionalOperatorNumberParam,
  value?: ConditionalOperator
};
