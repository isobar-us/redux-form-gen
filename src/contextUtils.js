// @flow
import React from 'react';
import PropTypes from 'prop-types';

import createReactContext from 'create-react-context';
export const GenContext = createReactContext({wasGenerated: false});

export const consumeGenContext = (Component: Function) => (props: mixed) => (
  <GenContext.Consumer>
    {(gen) => <Component {...props} gen={gen} /> }
  </GenContext.Consumer>
);

// inspired by
// https://medium.com/react-ecosystem/how-to-handle-react-context-a7592dfdcbc

// const provideContext = (childContextTypes, getChildContext) => (WrappedComponent) => {
//   class ContextProvider extends Component {
//     static childContextTypes = childContextTypes;
//     getChildContext = () => getChildContext(this.props);
//
//     render() {
//       return <WrappedComponent {...this.props} />;
//     }
//   }
//   return ContextProvider;
// };

const consumeContext = (contextTypes) => (Component: Function) => {
  /* The context is passed as props. This way the component is
   completely decoupled from the context API.
  */
  const ContextConsumer = (props: mixed, context: mixed) => <Component {...props} {...context} />;
  ContextConsumer.contextTypes = contextTypes;
  return ContextConsumer;
};

export const reduxFormContext = {
  _reduxForm: PropTypes.object
};

export const consumeReduxFormContext = consumeContext(reduxFormContext);
