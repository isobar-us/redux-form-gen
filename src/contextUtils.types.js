import type {Props as FormGeneratorProps} from './FormGenerator.types';

export type GenContextProps = {
  ...FormGeneratorProps,
  getCachedValue: Function,
  setCachedValue: Function
};
