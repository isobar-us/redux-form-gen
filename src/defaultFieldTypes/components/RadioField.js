import React, {Component} from 'react';
import cn from 'classnames';
import FieldMetaMessages from './FieldMetaMessages';
import Fragment from 'react-dot-fragment';

class RadioField extends Component {
  onChange = (e) => {
    let {input: {onChange}, handleChange} = this.props;

    if (handleChange) {
      handleChange(e.target.value);
    }

    onChange(e.target.value);
  };

  render() {
    let {input: {name, value: currentValue}, values, display, disabled} = this.props;

    let extraInputProps = Object.assign({}, this.props.input, {onChange: this.onChange});

    return (
      <Fragment>
        <div
          className={cn('wrapper', {
            'wrapper--stacked': display === 'stacked',
            'wrapper--inline': !display || display === 'inline'
          })}
        >
          {values.map(({label, value}, index) => {
            const id = `${name}-${value}`;
            return (
              <label key={index} htmlFor={id}>
                <input
                  key={index}
                  type='radio'
                  {...extraInputProps}
                  value={value}
                  id={id}
                  disabled={disabled}
                  checked={value === currentValue}
                />
                {label}
              </label>
            );
          })}
        </div>
        <FieldMetaMessages {...this.props.meta} />
      </Fragment>
    );
  }
}

export default RadioField;
