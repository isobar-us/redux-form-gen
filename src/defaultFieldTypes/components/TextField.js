import React, {Component} from 'react';
import FieldMetaMessages from './FieldMetaMessages';
import Fragment from 'react-dot-fragment';

class TextField extends Component {
  render() {
    const {input, disabled} = this.props;
    return (
      <Fragment>
        <input {...input} disabled={disabled} type='text' />
        <FieldMetaMessages {...this.props.meta} />
      </Fragment>
    );
  }
}

export default TextField;
