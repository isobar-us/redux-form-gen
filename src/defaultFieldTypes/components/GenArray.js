// import PropTypes from 'prop-types';
import React, {Component} from 'react';
import {FormSection} from 'redux-form';

import Fragment from 'react-dot-fragment';
import {consumeGenContext} from '../../contextUtils';
import GenField from '../../GenField';
import {getDefaultValues, getGenContextOptions} from '../../utils';
import isNil from 'lodash/isNil';

class GenArray extends Component {
  handleAddArrayItem = () => {
    const newItem = getDefaultValues({
      fields: [this.props.item],
      ...getGenContextOptions(this.props.gen),
      initialValues: {}
    });
    return this.props.fields.push(newItem);
  };

  isAddButtonEnabled = () => {
    const {fields, maxItems} = this.props;
    return isNil(maxItems) || fields.length < maxItems;
  };

  renderAddButtonLabel = () => {
    const {addLabel} = this.props;
    return !isNil(addLabel) ? addLabel : 'Add';
  };
  renderAddButton = () => {
    return (
      <button type='button' onClick={this.handleAddArrayItem}>
        {this.renderAddButtonLabel()}
      </button>
    );
  };

  render() {
    const {fields, item, itemProps} = this.props;
    return (
      <Fragment>
        {fields.map((name, index) => (
          <FormSection name={name} key={name}>
            <GenField {...{arrayItemProps: {array: fields, index}, field: item, path: `${itemProps.path}.item`}} />
          </FormSection>
        ))}
        {this.isAddButtonEnabled() && this.renderAddButton()}
      </Fragment>
    );
  }
}

const GenArrayUnrwrapped = GenArray;
export {GenArrayUnrwrapped};

export default consumeGenContext(GenArray);
