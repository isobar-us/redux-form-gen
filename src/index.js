// @flow
import FormGenerator from './FormGenerator';

import defaultFieldTypes, {getFieldOptions, genericFieldProps} from './defaultFieldTypes';
import GenericRequiredLabel from './defaultFieldTypes/components/GenericRequiredLabel';
import RequiredIndicator from './defaultFieldTypes/components/RequiredIndicator';
import injectGenProps from './injectGenProps';

import {
  isSectionEmpty,
  isSectionEmptyIterator,
  isSectionFilled,
  isSectionFilledIterator,
  getSectionErrors,
  getSectionErrorsIterator,
  isNilOrEmpty,
  getFieldPath,
  isFieldHidden,
  isFieldVisible,
  isFieldRequired,
  isFieldDisabled,
  isFieldEmpty,
  isFieldFilled,
  isFieldValid,
  mapFieldChildren
} from './validators';

import {getDefaultValues, buildLookupTable} from './utils';

import {consumeGenContext, GenContext} from './contextUtils';

import {evalCond, evalCondValid} from './conditionalUtils';

import GenField from './GenField';
import GenWrapper from './GenWrapper';

export type {FieldOptions, CustomFieldTypes} from './types';

export {
  FormGenerator,
  // defaultFieldTypes
  defaultFieldTypes,
  getFieldOptions,
  RequiredIndicator,
  GenericRequiredLabel,
  genericFieldProps,
  injectGenProps,
  // validators
  isSectionEmpty,
  isSectionEmptyIterator,
  isSectionFilled,
  isSectionFilledIterator,
  getSectionErrors,
  getSectionErrorsIterator,
  isNilOrEmpty,
  // validator utils
  getFieldPath,
  isFieldHidden,
  isFieldVisible,
  isFieldRequired,
  isFieldDisabled,
  isFieldEmpty,
  isFieldFilled,
  isFieldValid,
  mapFieldChildren,
  // utils
  getDefaultValues,
  buildLookupTable,
  // contextUtils
  consumeGenContext,
  GenContext,
  // conditionalUtils
  evalCond,
  evalCondValid,
  // internals
  GenField,
  GenWrapper
};
