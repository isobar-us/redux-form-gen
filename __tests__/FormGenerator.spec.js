import React from 'react';
import {mount} from 'enzyme';

import {reduxForm, reducer as formReducer, isDirty, getFormValues} from 'redux-form';
import {createStore, combineReducers} from 'redux';
import {Provider} from 'react-redux';

import {FormGenerator, getDefaultValues, defaultFieldTypes} from '../src';
import RadioField from '../src/defaultFieldTypes/components/RadioField';
import TextField from '../src/defaultFieldTypes/components/TextField';
import GenArray from '../src/defaultFieldTypes/components/GenArray';

const reducer = combineReducers({form: formReducer});
let store = createStore(reducer, {});

const Form = reduxForm({
  form: 'testForm'
})(({children}) => <div>{children}</div>);
const isDirtySelector = isDirty('testForm');
const getFormValuesSelector = getFormValues('testForm');

const FormDecorator = ({children, ...props}) => (
  <Provider store={store}>
    <Form {...props}>{children}</Form>
  </Provider>
);

// reset the redux store after each test
beforeEach(() => {
  store = createStore(reducer, {});
});

let consoleErrorSpy;
const startErrorSupression = () => {
  consoleErrorSpy = jest.spyOn(console, 'error');
  consoleErrorSpy.mockImplementation(() => {}); // suppress error log
};

const stopErrorSuppression = () => {
  consoleErrorSpy.mockReset();
  consoleErrorSpy.mockRestore();
};

describe('<FormGenerator/>', () => {
  it('should throw an error if rendered outside reduxForm', () => {
    startErrorSupression();
    expect(() => {
      mount(<FormGenerator fields={[]} />);
    }).toThrowErrorMatchingSnapshot();
    stopErrorSuppression();
  });

  it('should render null when no fields passed, and error on PropTypes', () => {
    startErrorSupression();
    mount(
      <FormDecorator>
        <FormGenerator />
      </FormDecorator>
    );
    expect(consoleErrorSpy).toHaveBeenCalledTimes(1);
    stopErrorSuppression();
  });

  it('should not render any fields', () => {
    const wrapper = mount(
      <FormDecorator>
        <FormGenerator fields={[]} />
      </FormDecorator>
    );
    expect(wrapper.find('label').exists()).toBe(false);
    expect(wrapper.find('input').exists()).toBe(false);
  });

  it('should not render _genHidden field types', () => {
    const fieldTypes = {
      hiddenText: ({field}) => ({
        ...defaultFieldTypes.text({field}),
        _genHidden: true
      })
    };
    const wrapper = mount(
      <FormDecorator>
        <FormGenerator
          fields={[
            {
              type: 'hiddenText',
              label: 'a label',
              questionId: 'foo'
            }
          ]}
          customFieldTypes={fieldTypes}
        />
      </FormDecorator>
    );

    expect(wrapper.find('label').exists()).toBe(false);
    expect(wrapper.find('input').exists()).toBe(false);
  });

  it('should render all fields disabled', () => {
    const wrapper = mount(
      <FormDecorator>
        <FormGenerator
          disabled
          fields={[
            {
              type: 'text',
              label: 'a label',
              questionId: 'foo'
            }
          ]}
        />
      </FormDecorator>
    );

    expect(wrapper.find('input').props().disabled).toBe(true);
  });

  it('should render all fields enabled', () => {
    const wrapper = mount(
      <FormDecorator>
        <FormGenerator
          fields={[
            {
              type: 'text',
              label: 'a label',
              questionId: 'foo'
            }
          ]}
        />
      </FormDecorator>
    );

    expect(wrapper.find('input').props().disabled).toBeUndefined();
  });

  it('should attach props to fields via customQuestionProps', () => {
    const customQuestionProps = {
      bar: {
        customProp: true
      }
    };
    const fields = [
      {
        type: 'text',
        questionId: 'bar'
      }
    ];
    const wrapper = mount(
      <FormDecorator>
        <FormGenerator fields={fields} customQuestionProps={customQuestionProps} />
      </FormDecorator>
    );

    expect(wrapper.find(TextField).props().customProp).toBe(true);
  });

  describe('type text', () => {
    it('should render field and label', () => {
      const wrapper = mount(
        <FormDecorator>
          <FormGenerator fields={[{type: 'text', questionId: 'foo', label: 'Foo'}]} />
        </FormDecorator>
      );
      expect(wrapper.find('label').text()).toBe('Foo');
      expect(wrapper.find('input').exists()).toBe(true);
    });

    it('should not show hidden fields', () => {
      const wrapper = mount(
        <FormDecorator>
          <FormGenerator
            fields={[
              {
                type: 'text',
                questionId: 'foo',
                label: 'Foo',
                childFields: [
                  {
                    type: 'text',
                    questionId: 'bar',
                    label: 'Bar',
                    conditionalVisible: {
                      value: '123'
                    }
                  }
                ]
              }
            ]}
          />
        </FormDecorator>
      );
      expect(wrapper.find('.section--hidden').length).toBe(1);
    });

    it('should show hidden fields if cond matches', () => {
      const wrapper = mount(
        <FormDecorator initialValues={{foo: '123'}}>
          <FormGenerator
            fields={[
              {
                type: 'text',
                questionId: 'foo',
                label: 'Foo',
                childFields: [
                  {
                    type: 'text',
                    questionId: 'bar',
                    label: 'Bar',
                    conditionalVisible: {
                      value: '123'
                    }
                  }
                ]
              }
            ]}
          />
        </FormDecorator>
      );
      expect(wrapper.find('.wrapper--hidden').length).toBe(0);
    });

    it('should be disabled if cond matches', () => {
      const wrapper = mount(
        <FormDecorator initialValues={{foo: '123'}}>
          <FormGenerator
            fields={[
              {
                type: 'text',
                questionId: 'foo',
                label: 'Foo',
                childFields: [
                  {
                    type: 'text',
                    questionId: 'bar',
                    label: 'Bar',
                    conditionalDisabled: {
                      value: '123'
                    }
                  }
                ]
              }
            ]}
          />
        </FormDecorator>
      );
      const foo = wrapper.findWhere((n) => n.is(TextField) && n.props().input.name === 'foo');
      const bar = wrapper.findWhere((n) => n.is(TextField) && n.props().input.name === 'bar');
      expect(foo.props().disabled).toBeUndefined();
      expect(bar.props().disabled).toBe(true);
    });

    it('should apply display = inline style from gen props', () => {
      const wrapper = mount(
        <FormDecorator initialValues={{foo: '123'}}>
          <FormGenerator
            fields={[
              {
                type: 'text',
                questionId: 'foo',
                label: 'Foo'
              }
            ]}
            display='inline'
          />
        </FormDecorator>
      );

      expect(wrapper.find('.wrapper--inline').length).toBe(1);
    });

    it('should default to display = stacked', () => {
      const wrapper = mount(
        <FormDecorator initialValues={{foo: '123'}}>
          <FormGenerator
            fields={[
              {
                type: 'text',
                questionId: 'foo',
                label: 'Foo'
              }
            ]}
          />
        </FormDecorator>
      );

      expect(wrapper.find('.wrapper--stacked').length).toBe(1);
    });
  });

  describe('dirty/prisine with CondClearField', () => {
    const conditionalStructure = [
      {
        type: 'radio',
        label: 'one',
        questionId: 'one',
        options: [{label: 'Yes', value: 'yes'}, {label: 'No', value: 'no'}],
        childFields: [
          {
            type: 'array',
            questionId: 'array',
            conditionalVisible: {value: 'yes'},
            item: {
              type: 'arrayItem',
              childFields: [
                {
                  type: 'array',
                  questionId: 'subArray',
                  defaultValue: [
                    {
                      subArray_item_text: ''
                    }
                  ],
                  item: {
                    type: 'arrayItem',
                    childFields: [
                      {
                        type: 'text',
                        questionId: 'subArray_item_text'
                      }
                    ]
                  }
                },
                {
                  type: 'radio',
                  label: 'one',
                  questionId: 'array_item_vis',
                  options: [{label: 'Yes', value: 'yes'}, {label: 'No', value: 'no'}]
                },
                {
                  type: 'text',
                  questionId: 'array_item_text',
                  conditionalVisible: {
                    questionId: 'array_item_vis',
                    equals: 'yes'
                  }
                }
              ]
            }
          },
          {
            type: 'radio',
            label: 'two',
            questionId: 'two',
            conditionalVisible: {value: 'yes'},
            options: [{label: 'Yes', value: 'yes'}, {label: 'No', value: 'no'}],
            childFields: [
              {
                type: 'radio',
                label: 'three',
                questionId: 'three',
                conditionalVisible: {value: 'yes'},
                options: [{label: 'Yes', value: 'yes'}, {label: 'No', value: 'no'}],
                childFields: [
                  {
                    type: 'radio',
                    label: 'four',
                    questionId: 'four',
                    conditionalVisible: {value: 'yes'},
                    options: [{label: 'Yes', value: 'yes'}, {label: 'No', value: 'no'}]
                  }
                ]
              }
            ]
          }
        ]
      }
    ];

    it('should remove & restore hidden values from the state', () => {
      const wrapper = mount(
        <FormDecorator
          initialValues={getDefaultValues({
            fields: conditionalStructure,
            initialValues: {
              one: 'no'
            }
          })}
        >
          <FormGenerator fields={conditionalStructure} />
        </FormDecorator>
      );
      const one = wrapper.findWhere((n) => n.is(RadioField) && n.props().input.name === 'one');
      const two = wrapper.findWhere((n) => n.is(RadioField) && n.props().input.name === 'two');
      const three = wrapper.findWhere((n) => n.is(RadioField) && n.props().input.name === 'three');
      const four = wrapper.findWhere((n) => n.is(RadioField) && n.props().input.name === 'four');
      const array = wrapper.findWhere((n) => n.is(GenArray) && n.props().fields.name === 'array');

      one.props().input.onChange('yes');
      two.props().input.onChange('yes');
      three.props().input.onChange('yes');
      four.props().input.onChange('no');
      array.props().fields.push({
        array_item_vis: 'yes',
        array_item_text: 'something'
      });
      array.props().fields.push({
        array_item_vis: 'yes',
        array_item_text: 'another'
      });

      let state = store.getState();
      expect(isDirtySelector(state)).toBe(true);
      expect(getFormValuesSelector(state)).toEqual({
        one: 'yes',
        two: 'yes',
        three: 'yes',
        four: 'no',
        array: [
          {
            array_item_vis: 'yes',
            array_item_text: 'something'
          },
          {
            array_item_vis: 'yes',
            array_item_text: 'another'
          }
        ]
      });

      one.props().input.onChange('no');
      state = store.getState();
      expect(isDirtySelector(state)).toBe(false);
      expect(getFormValuesSelector(state)).toEqual({
        one: 'no',
        array: []
      });

      one.props().input.onChange('yes');
      state = store.getState();
      expect(isDirtySelector(state)).toBe(true);
      expect(getFormValuesSelector(state)).toEqual({
        one: 'yes',
        two: 'yes',
        three: 'yes',
        four: 'no',
        array: [
          {
            array_item_vis: 'yes',
            array_item_text: 'something'
          },
          {
            array_item_vis: 'yes',
            array_item_text: 'another'
          }
        ]
      });
    });

    it('should allow reset of cachedValues via FormGenerator.clearCachedValues()', () => {
      const wrapper = mount(
        <FormDecorator
          initialValues={getDefaultValues({
            fields: conditionalStructure,
            initialValues: {
              one: 'yes',
              two: 'yes',
              three: 'yes',
              four: 'no',
              array: [
                {
                  array_item_text: 'something'
                }
              ]
            }
          })}
        >
          <FormGenerator fields={conditionalStructure} />
        </FormDecorator>
      );

      const one = wrapper.findWhere((n) => n.is(RadioField) && n.props().input.name === 'one');

      let state = store.getState();
      one.props().input.onChange('no');

      state = store.getState();
      expect(isDirtySelector(state)).toBe(true);
      expect(getFormValuesSelector(state)).toEqual({
        one: 'no',
        two: '',
        three: '',
        four: '',
        array: []
      });

      wrapper
        .find('FormGenerator')
        .instance()
        .clearCachedValues();

      one.props().input.onChange('yes');
      state = store.getState();
      expect(isDirtySelector(state)).toBe(true);
      expect(getFormValuesSelector(state)).toEqual({
        one: 'yes',
        two: '',
        three: '',
        four: '',
        array: []
      });
    });

    it('should remove & restore hidden initialValues', () => {
      const wrapper = mount(
        <FormDecorator
          initialValues={getDefaultValues({
            fields: conditionalStructure,
            initialValues: {
              one: 'yes',
              two: 'yes',
              three: 'yes',
              four: 'no',
              array: [
                {
                  array_item_text: 'something'
                }
              ]
            }
          })}
        >
          <FormGenerator fields={conditionalStructure} />
        </FormDecorator>
      );
      const one = wrapper.findWhere((n) => n.is(RadioField) && n.props().input.name === 'one');

      let state = store.getState();
      expect(isDirtySelector(state)).toBe(false);
      expect(getFormValuesSelector(state)).toEqual({
        one: 'yes',
        two: 'yes',
        three: 'yes',
        four: 'no',
        array: [
          {
            subArray: [
              {
                subArray_item_text: ''
              }
            ],
            array_item_text: 'something'
          }
        ]
      });

      one.props().input.onChange('no');
      state = store.getState();
      expect(isDirtySelector(state)).toBe(true);
      expect(getFormValuesSelector(state)).toEqual({
        one: 'no',
        two: '',
        three: '',
        four: '',
        array: []
      });

      one.props().input.onChange('yes');
      state = store.getState();
      expect(isDirtySelector(state)).toBe(false);
      expect(getFormValuesSelector(state)).toEqual({
        one: 'yes',
        two: 'yes',
        three: 'yes',
        four: 'no',
        array: [
          {
            subArray: [
              {
                subArray_item_text: ''
              }
            ],
            array_item_text: 'something'
          }
        ]
      });
    });
  });
});
