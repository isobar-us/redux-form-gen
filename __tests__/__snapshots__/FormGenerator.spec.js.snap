// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormGenerator/> should throw an error if rendered outside reduxForm 1`] = `"FormGenerator must be inside a component decorated with reduxForm()"`;
